Updated Bin transform.

jheer opened the pull request:
This PR updates the bin transform in two ways:
 1. Perform automatic calculation of min/max if not specified.
 2. Output both bin start and end values to simplify later operations.

Item 1 is implemented directly using datalib's `extent` method rather than a streaming aggregator. To allow this, `Bin` is now a `BatchTransform` instance. This seems simpler and lower overhead than instantiating an aggregator and all its associated state (which would include another full array with all the tuples in it). Simple min/max scans should be pretty fast, and can also be cached more effectively in future updates.

Item 2 extends the binning output with an `"end"` parameter that defaults to `"bin_end"`. For internal consistency, it would be nice to have `"start"` and `"end"` parameters that default to `"bin_start"` and `"bin_end"`. However, for backwards-compatibility I've left the bin_start parameter as simply `"bin"`. Any thoughts on whether we should make a breaking change here? It appears that the crossfilter example is currently the only public example that uses the bin transform, so that is what we would need to update, but any changes here could adversely impact end users.

jheer commented:
Here is an example spec for a histogram annotated with median and quartile boundaries, demonstrating how the new bin start/end outputs simplify scale construction and mark composition, all along a shared numerical axis. @domoritz @kanitw 

```
{
  "width": 400,
  "height": 200,
  "padding": {"top": 10, "left": 30, "bottom": 30, "right": 10},
  "data": [
    {
      "name": "source",
      "values": [
        {"x": 1,  "y": 28}, {"x": 2,  "y": 55},
        {"x": 3,  "y": 43}, {"x": 4,  "y": 91},
        {"x": 5,  "y": 81}, {"x": 6,  "y": 53},
        {"x": 7,  "y": 19}, {"x": 8,  "y": 87},
        {"x": 9,  "y": 52}, {"x": 10, "y": 48},
        {"x": 11, "y": 24}, {"x": 12, "y": 49},
        {"x": 13, "y": 87}, {"x": 14, "y": 66},
        {"x": 15, "y": 17}, {"x": 16, "y": 27},
        {"x": 17, "y": 68}, {"x": 18, "y": 16},
        {"x": 19, "y": 49}, {"x": 20, "y": 15}
      ],
      "transform": [
        { "type": "bin", "field": "y", "maxbins": 10,
          "output": {"bin": "bin_start", "end": "bin_end"} }
      ]
    },
    {
      "name": "bins",
      "source": "source",
      "transform": [
        { "type": "aggregate", "groupby": ["bin_start", "bin_end"],
          "summarize": {"*": "count"} }
      ]
    },
    {
      "name": "stats",
      "source": "source",
      "transform": [
        { "type": "aggregate",
          "summarize": {"y": ["q1", "median", "q3"]} }
      ]
    }
  ],
  "scales": [
    {
      "name": "x",
      "type": "linear",
      "range": "width",
      "zero": false, "round": true,
      "domain": {"data": "bins", "field": ["bin_start", "bin_end"]}
    },
    {
      "name": "y",
      "type": "linear",
      "range": "height",
      "domain": {"data": "bins", "field": "count"},
      "nice": true
    }
  ],
  "axes": [
    {"type": "x", "scale": "x"},
    {"type": "y", "scale": "y"}
  ],
  "marks": [
    {
      "type": "rect",
      "from": {"data": "stats"},
      "properties": {
        "enter": {
          "x": {"scale": "x", "field": "q1_y"},
          "x2": {"scale": "x", "field": "median_y"},
          "y": {"value": -5},
          "height": {"field": {"group": "height"}, "offset": 5},
          "fill": {"value": "lightgray"},
          "fillOpacity": {"value": 0.6}
        }
      }
    },
    {
      "type": "rect",
      "from": {"data": "stats"},
      "properties": {
        "enter": {
          "x": {"scale": "x", "field": "median_y"},
          "x2": {"scale": "x", "field": "q3_y"},
          "y": {"value": -5},
          "height": {"field": {"group": "height"}, "offset": 5},
          "fill": {"value": "lightgray"}
        }
      }
    },
    {
      "type": "rect",
      "from": {"data": "bins"},
      "properties": {
        "enter": {
          "x": {"scale": "x", "field": "bin_start"},
          "x2": {"scale": "x", "field": "bin_end", "offset": -1},
          "y": {"scale": "y", "field": "count"},
          "y2": {"scale": "y", "value": 0}
        },
        "update": {
          "fill": {"value": "steelblue"}
        },
        "hover": {
          "fill": {"value": "red"}
        }
      }
    },
    {
      "type": "rect",
      "from": {"data": "stats"},
      "properties": {
        "enter": {
          "x": {"scale": "x", "field": "median_y"},
          "width": {"value": 2},
          "y": {"value": -5},
          "height": {"field": {"group": "height"}, "offset": 5},
          "fill": {"value": "crimson"}
        }
      }
    }
  ]
}
```

arvind commented:
My vote would be to rename `bin` to `start`. We can bump the minor version and message the release notes to call attention to and mitigate adverse effects to end-users.

jheer commented:
I agree. I think we constitute a quorum :) I will make the change and update the PR.

kanitw commented:
+1 on rename `bin` to `start`.

jheer commented:
OK, updated bin outputs and updated crossfilter test spec. The crossfilter example in vega-editor will also need to be updated once this is merged and released. I will now update the above comment with the working example spec. We also need to remember to update the documentation upon release.

jeffbaumes commented:
:clap: I had run into this issue too on building a generic histogram component, thanks for fixing!
